Email notification scheduled job throws NullPointerException exception

ERROR [20-January 06:03:05] [schedulerFactoryBean_Worker-3] org.quartz.core.JobRunShell  - Job EMAIL.TASK_SHOULD_START threw an unhandled Exception:
java.lang.NullPointerException
        at org.libreplan.importers.notifications.ComposeMessage.composeMessageForUser(ComposeMessage.java:138)
        at org.libreplan.importers.notifications.realization.SendEmailOnTaskShouldStart.composeMessageForUser(SendEmailOnTaskShouldStart.java:101)
        at org.libreplan.importers.notifications.realization.SendEmailOnTaskShouldStart.sendEmail(SendEmailOnTaskShouldStart.java:92)
        at sun.reflect.GeneratedMethodAccessor187.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
        at com.sun.proxy.$Proxy217.sendEmail(Unknown Source)
        at org.libreplan.importers.notifications.jobs.SendEmailOnTaskShouldStartJob.executeInternal(SendEmailOnTaskShouldStartJob.java:48)
        at org.springframework.scheduling.quartz.QuartzJobBean.execute(QuartzJobBean.java:75)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
ERROR [20-January 06:03:05] [schedulerFactoryBean_Worker-3] org.quartz.core.ErrorLogger  - Job (EMAIL.TASK_SHOULD_START threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: java.lang.NullPointerException]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.lang.NullPointerException
        at org.libreplan.importers.notifications.ComposeMessage.composeMessageForUser(ComposeMessage.java:138)
        at org.libreplan.importers.notifications.realization.SendEmailOnTaskShouldStart.composeMessageForUser(SendEmailOnTaskShouldStart.java:101)
        at org.libreplan.importers.notifications.realization.SendEmailOnTaskShouldStart.sendEmail(SendEmailOnTaskShouldStart.java:92)
        at sun.reflect.GeneratedMethodAccessor187.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
        at com.sun.proxy.$Proxy217.sendEmail(Unknown Source)
        at org.libreplan.importers.notifications.jobs.SendEmailOnTaskShouldStartJob.executeInternal(SendEmailOnTaskShouldStartJob.java:48)
        at org.springframework.scheduling.quartz.QuartzJobBean.execute(QuartzJobBean.java:75)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        ... 1 more

--- a/libreplan-business/src/main/java/org/libreplan/business/resources/daos/IWorkerDAO.java
+++ b/libreplan-business/src/main/java/org/libreplan/business/resources/daos/IWorkerDAO.java
@@ -122,8 +122,11 @@ public interface IWorkerDAO extends IIntegrationEntityDAO<Worker> {
     public List<Worker> findByFirstNameSecondNameAnotherTransaction(
             String firstname, String secondname);
 
+    public Worker getCurrentWorker(Long resourceID);
+
     /**
      * Return the list of {@link Worker Workers} bound to any {@link User}.
      */
     List<Worker> getBound();
+
 }

--- a/libreplan-business/src/main/java/org/libreplan/business/resources/daos/WorkerDAO.java
+++ b/libreplan-business/src/main/java/org/libreplan/business/resources/daos/WorkerDAO.java
@@ -198,4 +198,18 @@ public class WorkerDAO extends IntegrationEntityDAO<Worker>
         criteria.add(Restrictions.isNotNull("user"));
         return criteria.list();
     }
+
+    @Override
+    @Transactional(readOnly = true)
+    public Worker getCurrentWorker(Long resourceID) {
+        List<Worker> workerList = getWorkers();
+
+        for (Worker worker : workerList) {
+            if (worker.getId().equals(resourceID)) {
+                return worker;
+            }
+        }
+
+        return null;
+    }
 }

--- a/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/ComposeMessage.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/ComposeMessage.java
@@ -27,6 +27,7 @@ import org.libreplan.business.common.entities.ConnectorProperty;
 import org.libreplan.business.email.entities.EmailNotification;
 import org.libreplan.business.email.entities.EmailTemplate;
 import org.libreplan.business.email.entities.EmailTemplateEnum;
+import org.libreplan.business.resources.daos.IWorkerDAO;
 import org.libreplan.business.resources.entities.Resource;
 import org.libreplan.business.resources.entities.Worker;
 import org.libreplan.business.settings.entities.Language;
@@ -102,7 +103,7 @@ import static org.libreplan.web.I18nHelper._;
 public class ComposeMessage {
 
 	@Autowired
-	private IWorkerModel workerModel;
+	private IWorkerDAO workerDAO;
 
 	@Autowired
 	private IEmailTemplateModel emailTemplateModel;
@@ -131,11 +132,11 @@ public class ComposeMessage {
 		Resource resource = notification.getResource();
 		EmailTemplateEnum type = notification.getType();
 		Locale locale;
-		Worker currentWorker = getCurrentWorker(resource.getId());
+		Worker currentWorker = workerDAO.getCurrentWorker(resource.getId());
 
 		UserRole currentUserRole = getCurrentUserRole(notification.getType());
 
-		if (currentWorker != null && currentWorker.getUser().isInRole(currentUserRole)) {
+		if (currentWorker != null && (currentWorker.getUser() != null) && currentWorker.getUser().isInRole(currentUserRole)) {
 			if (currentWorker.getUser().getApplicationLanguage().equals(Language.BROWSER_LANGUAGE)) {
 				locale = new Locale(System.getProperty("user.language"));
 			} else {
@@ -195,18 +196,6 @@ public class ComposeMessage {
 			}
 		}
 		return false;
-	}
-
-	private Worker getCurrentWorker(Long resourceID) {
-		List<Worker> workerList = workerModel.getWorkers();
-
-		for (Worker aWorkerList : workerList) {
-			if (aWorkerList.getId().equals(resourceID)) {
-				return aWorkerList;
-			}
-		}
-
-		return null;
 	}
 
 	private EmailTemplate findCurrentEmailTemplate(EmailTemplateEnum templateEnum, Locale locale) {

--- a/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/realization/SendEmailOnMilestoneReached.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/realization/SendEmailOnMilestoneReached.java
@@ -112,10 +112,6 @@ public class SendEmailOnMilestoneReached implements IEmailNotificationJob {
     }
 
     private void sendEmailNotificationToManager(TaskElement item) {
-        emailNotificationModel.setNewObject();
-        emailNotificationModel.setType(EmailTemplateEnum.TEMPLATE_MILESTONE_REACHED);
-        emailNotificationModel.setUpdated(new Date());
-
         String responsible = "";
         if ( item.getParent().getOrderElement().getOrder().getResponsible() != null ) {
             responsible = item.getParent().getOrderElement().getOrder().getResponsible();
@@ -129,6 +125,9 @@ public class SendEmailOnMilestoneReached implements IEmailNotificationJob {
                 user.isInRole(UserRole.ROLE_SUPERUSER) || user.isInRole(UserRole.ROLE_EMAIL_MILESTONE_REACHED);
 
 	        if ( user.getWorker() != null && userHasNeededRoles ) {
+	            emailNotificationModel.setNewObject();
+	            emailNotificationModel.setType(EmailTemplateEnum.TEMPLATE_MILESTONE_REACHED);
+	            emailNotificationModel.setUpdated(new Date());
 	            emailNotificationModel.setResource(user.getWorker());
 	            emailNotificationModel.setTask(item);
 	            emailNotificationModel.setProject(item.getParent());

--- a/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/realization/SendEmailOnTaskShouldFinish.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/realization/SendEmailOnTaskShouldFinish.java
@@ -28,11 +28,15 @@ import org.libreplan.business.email.entities.EmailTemplateEnum;
 import org.libreplan.business.planner.daos.ITaskElementDAO;
 import org.libreplan.business.planner.entities.ResourceAllocation;
 import org.libreplan.business.planner.entities.TaskElement;
+import org.libreplan.business.resources.daos.IWorkerDAO;
 import org.libreplan.business.resources.entities.Resource;
+import org.libreplan.business.resources.entities.Worker;
+import org.libreplan.business.users.entities.UserRole;
 import org.libreplan.importers.notifications.ComposeMessage;
 import org.libreplan.importers.notifications.EmailConnectionValidator;
 import org.libreplan.importers.notifications.IEmailNotificationJob;
 import org.libreplan.web.email.IEmailNotificationModel;
+import org.libreplan.web.resources.worker.IWorkerModel;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.config.BeanDefinition;
 import org.springframework.context.annotation.Scope;
@@ -67,6 +71,9 @@ public class SendEmailOnTaskShouldFinish implements IEmailNotificationJob {
 
     @Autowired
     private EmailConnectionValidator emailConnectionValidator;
+
+    @Autowired
+    private IWorkerDAO workerDAO;
 
     /**
      * Transactional here is needed because without this annotation we are getting
@@ -122,7 +129,6 @@ public class SendEmailOnTaskShouldFinish implements IEmailNotificationJob {
         }
     }
 
-
     private void sendEmailNotificationAboutTaskShouldFinish(TaskElement item){
         List<ResourceAllocation<?>> resourceAllocations = new ArrayList<>(item.getAllResourceAllocations());
 
@@ -131,13 +137,17 @@ public class SendEmailOnTaskShouldFinish implements IEmailNotificationJob {
             resources.add(allocation.getAssociatedResources().get(0));
 
         for (Resource resourceItem : resources){
-            emailNotificationModel.setNewObject();
-            emailNotificationModel.setType(EmailTemplateEnum.TEMPLATE_TODAY_TASK_SHOULD_FINISH);
-            emailNotificationModel.setUpdated(new Date());
-            emailNotificationModel.setResource(resourceItem);
-            emailNotificationModel.setTask(item);
-            emailNotificationModel.setProject(item.getParent());
-            emailNotificationModel.confirmSave();
+            Worker currentWorker = workerDAO.getCurrentWorker(resourceItem.getId());
+
+            if (currentWorker != null && (currentWorker.getUser() != null) && currentWorker.getUser().isInRole(UserRole.ROLE_EMAIL_TASK_SHOULD_FINISH)) {
+                emailNotificationModel.setNewObject();
+                emailNotificationModel.setType(EmailTemplateEnum.TEMPLATE_TODAY_TASK_SHOULD_FINISH);
+                emailNotificationModel.setUpdated(new Date());
+                emailNotificationModel.setResource(resourceItem);
+                emailNotificationModel.setTask(item);
+                emailNotificationModel.setProject(item.getParent());
+                emailNotificationModel.confirmSave();
+            }
         }
     }
 

--- a/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/realization/SendEmailOnTaskShouldStart.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/realization/SendEmailOnTaskShouldStart.java
@@ -28,16 +28,22 @@ import org.libreplan.business.email.entities.EmailTemplateEnum;
 import org.libreplan.business.planner.daos.ITaskElementDAO;
 import org.libreplan.business.planner.entities.ResourceAllocation;
 import org.libreplan.business.planner.entities.TaskElement;
+import org.libreplan.business.resources.daos.IWorkerDAO;
 import org.libreplan.business.resources.entities.Resource;
+import org.libreplan.business.resources.entities.Worker;
+import org.libreplan.business.users.entities.User;
+import org.libreplan.business.users.entities.UserRole;
 import org.libreplan.importers.notifications.ComposeMessage;
 import org.libreplan.importers.notifications.EmailConnectionValidator;
 import org.libreplan.importers.notifications.IEmailNotificationJob;
 import org.libreplan.web.email.IEmailNotificationModel;
+import org.libreplan.web.resources.worker.IWorkerModel;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.config.BeanDefinition;
 import org.springframework.context.annotation.Scope;
 import org.springframework.stereotype.Component;
 import org.springframework.transaction.annotation.Transactional;
+import org.zkoss.zkplus.spring.SpringUtil;
 
 import java.util.ArrayList;
 import java.util.Date;
@@ -67,6 +73,9 @@ public class SendEmailOnTaskShouldStart implements IEmailNotificationJob {
 
     @Autowired
     private EmailConnectionValidator emailConnectionValidator;
+
+    @Autowired
+    private IWorkerDAO workerDAO;
 
     /**
      * Transactional here is needed because without this annotation we are getting
@@ -130,13 +139,17 @@ public class SendEmailOnTaskShouldStart implements IEmailNotificationJob {
             resources.add(allocation.getAssociatedResources().get(0));
 
         for (Resource resourceItem : resources) {
-            emailNotificationModel.setNewObject();
-            emailNotificationModel.setType(EmailTemplateEnum.TEMPLATE_TODAY_TASK_SHOULD_START);
-            emailNotificationModel.setUpdated(new Date());
-            emailNotificationModel.setResource(resourceItem);
-            emailNotificationModel.setTask(item);
-            emailNotificationModel.setProject(item.getParent());
-            emailNotificationModel.confirmSave();
+            Worker currentWorker = workerDAO.getCurrentWorker(resourceItem.getId());
+
+            if (currentWorker != null && (currentWorker.getUser() != null) && currentWorker.getUser().isInRole(UserRole.ROLE_EMAIL_TASK_SHOULD_START)) {
+                emailNotificationModel.setNewObject();
+                emailNotificationModel.setType(EmailTemplateEnum.TEMPLATE_TODAY_TASK_SHOULD_START);
+                emailNotificationModel.setUpdated(new Date());
+                emailNotificationModel.setResource(resourceItem);
+                emailNotificationModel.setTask(item);
+                emailNotificationModel.setProject(item.getParent());
+                emailNotificationModel.confirmSave();
+            }
         }
     }
 

--- a/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/realization/SendEmailOnTimesheetDataMissing.java
+++ b/libreplan-webapp/src/main/java/org/libreplan/importers/notifications/realization/SendEmailOnTimesheetDataMissing.java
@@ -33,6 +33,7 @@ import org.libreplan.business.resources.entities.Resource;
 import org.libreplan.business.resources.entities.Worker;
 
 import org.libreplan.business.users.entities.User;
+import org.libreplan.business.users.entities.UserRole;
 import org.libreplan.business.workingday.EffortDuration;
 import org.libreplan.business.workingday.IntraDayDate;
 import org.libreplan.business.workreports.daos.IWorkReportDAO;
@@ -159,11 +160,13 @@ public class SendEmailOnTimesheetDataMissing implements IEmailNotificationJob {
 
     private void addRowsToNotificationTable(List<User> users){
         for (User user : users){
-            emailNotificationModel.setNewObject();
-            emailNotificationModel.setResource(user.getWorker());
-            emailNotificationModel.setType(EmailTemplateEnum.TEMPLATE_ENTER_DATA_IN_TIMESHEET);
-            emailNotificationModel.setUpdated(new Date());
-            emailNotificationModel.confirmSave();
+            if ( user.isInRole(UserRole.ROLE_EMAIL_TIMESHEET_DATA_MISSING) ) {
+                emailNotificationModel.setNewObject();
+                emailNotificationModel.setResource(user.getWorker());
+                emailNotificationModel.setType(EmailTemplateEnum.TEMPLATE_ENTER_DATA_IN_TIMESHEET);
+                emailNotificationModel.setUpdated(new Date());
+                emailNotificationModel.confirmSave();
+            }
         }
     }
 

